stdlib/public/Reflection/TypeLowering.cpp

```diff
@@ -286,132 +287,737 @@ bool RecordTypeInfo::readExtraInhabitantIndex(remote::MemoryReader &reader,
+    andNotMask(mask.mask, mask.size, offset);
+  }
+
+  // Zero all bits except for the `n` most significant ones.
+  // XXX TODO: Big-endian support?
+  void keepOnlyMostSignificantBits(int n) {
+    int count = 0;
+    if (size < 1) {
+      return;
+    }
+    unsigned i = size;
+    while (i > 0) {
+      i -= 1;
+      if (count < n) {
+        for (int b = 128; b > 0; b >>= 1) {
+          if (count >= n) {
+            mask[i] &= ~b;
+          } else if ((mask[i] & b) != 0) {
+            ++count;
+          }
         }
       } else {
-        NonPayloadCaseCount += 1;
+        mask[i] = 0;
       }
     }
-    if (getSize() > PayloadSize) {
-      // Multipayload enums that do use a separate tag
-      // export XIs from that tag.
-      unsigned tag = 0;
-      auto TagSize = getSize() - PayloadSize;
-      auto TagAddress = remote::RemoteAddress(address.getAddressData() + PayloadSize);
-      if (!reader.readInteger(TagAddress, TagSize, &tag))
+  }
+
+  int numBits() const {
+    return size * 8;
+  }
+
+  int numSetBits() const {
+    int count = 0;
+    for (unsigned i = 0; i < size; ++i) {
+      if (mask[i] != 0) {
+        for (int b = 1; b < 256; b <<= 1) {
+          if ((mask[i] & b) != 0) {
+            ++count;
+          }
+        }
+      }
+    }
+    return count;
+  }
+
+  // Read a mask-sized area from the target and collect
+  // the masked bits into a single integer.
+  template<typename IntegerType>
+  bool readMaskedInteger(remote::MemoryReader &reader,
+                         remote::RemoteAddress address,
+                         IntegerType *dest) const {
+    auto data = reader.readBytes(address, size);
+    if (!data) {
+      return false;
+    }
+#if defined(__BIG_ENDIAN__)
+    assert(false && "Big endian not supported for readMaskedInteger");
+#else
+    IntegerType result = 0;
+    IntegerType resultBit = 1; // Start from least-significant bit
+    auto bytes = static_cast<const uint8_t *>(data.get());
+    for (unsigned i = 0; i < size; ++i) {
+      for (int b = 1; b < 256; b <<= 1) {
+        if ((mask[i] & b) != 0) {
+          if ((bytes[i] & b) != 0) {
+            result |= resultBit;
+          }
+          resultBit <<= 1;
+        }
+      }
+    }
+    *dest = result;
+    return true;
+#endif
+  }
+
+private:
+  void andMask(void *maskData, unsigned len, unsigned offset) {
+    assert(offset + len <= size);
+    uint8_t *maskBytes = (uint8_t *)maskData;
+    for (unsigned i = 0; i < len; ++i) {
+      mask[i + offset] &= maskBytes[i];
+    }
+  }
+
+  void andNotMask(void *maskData, unsigned len, unsigned offset) {
+    assert(offset + len <= size);
+    uint8_t *maskBytes = (uint8_t *)maskData;
+    for (unsigned i = 0; i < len; ++i) {
+      mask[i + offset] &= ~maskBytes[i];
+    }
+  }
+};
+
+// General multi-payload enum support for enums that do use spare
+// bits in the payload.
+class MultiPayloadEnumTypeInfo: public EnumTypeInfo {
```

Comment:
MP enums that do not use spare bits can rely on the much simpler `SimpleMultiPayloadEnumTypeInfo` above.